File: backend/app/geo.py
from math import atan2, cos, radians, sin, sqrt

EARTH_MI = 3958.8
DEFAULT_LAT = 40.4406
DEFAULT_LNG = -79.9959
PITTSBURGH_LAT = DEFAULT_LAT
PITTSBURGH_LNG = DEFAULT_LNG


def haversine_mi(lat1: float, lng1: float, lat2: float, lng2: float) -> float:
    dlat = radians(lat2 - lat1)
    dlng = radians(lng2 - lng1)
    a = sin(dlat / 2) ** 2 + cos(radians(lat1)) * cos(radians(lat2)) * sin(dlng / 2) ** 2
    return EARTH_MI * 2 * atan2(sqrt(a), sqrt(1 - a))


def within_radius(
    origin_lat: float,
    origin_lng: float,
    lat: float,
    lng: float,
    radius_mi: float,
) -> bool:
    return haversine_mi(origin_lat, origin_lng, lat, lng) <= radius_mi

File: frontend/src/lib/geo.ts
import type { Origin } from "@/lib/types";

export const PITTSBURGH: Origin = { lat: 40.4406, lng: -79.9959 };

export function haversineMi(a: Origin, b: Origin): number {
  const r = 3958.8;
  const dLat = ((b.lat - a.lat) * Math.PI) / 180;
  const dLng = ((b.lng - a.lng) * Math.PI) / 180;
  const x =
    Math.sin(dLat / 2) ** 2 +
    Math.cos((a.lat * Math.PI) / 180) * Math.cos((b.lat * Math.PI) / 180) * Math.sin(dLng / 2) ** 2;
  return r * 2 * Math.atan2(Math.sqrt(x), Math.sqrt(1 - x));
}

export function resolveOrigin(coords: Origin | null): { origin: Origin; outsideCity: boolean } {
  if (!coords) return { origin: PITTSBURGH, outsideCity: false };
  const outsideCity = haversineMi(coords, PITTSBURGH) > 25;
  return { origin: outsideCity ? PITTSBURGH : coords, outsideCity };
}

export function formatWhen(iso: string): string {
  return new Date(iso).toLocaleString(undefined, {
    weekday: "short",
    month: "short",
    day: "numeric",
    hour: "numeric",
    minute: "2-digit",
  });
}

export function toLocalInput(date: Date): string {
  const pad = (n: number) => String(n).padStart(2, "0");
  return `${date.getFullYear()}-${pad(date.getMonth() + 1)}-${pad(date.getDate())}T${pad(date.getHours())}:${pad(date.getMinutes())}`;
}

export function defaultWindow(): { start: string; end: string } {
  const start = new Date();
  const end = new Date(start.getTime() + 14 * 24 * 60 * 60 * 1000);
  return { start: toLocalInput(start), end: toLocalInput(end) };
}

export function queryParams(origin: Origin, radius: number, start: string, end: string, q = ""): URLSearchParams {
  const params = new URLSearchParams({
    lat: String(origin.lat),
    lng: String(origin.lng),
    radius_mi: String(radius),
    start: new Date(start).toISOString(),
    end: new Date(end).toISOString(),
  });
  if (q.trim()) params.set("q", q.trim());
  return params;
}
